### DISCUSSION TITLE
Unable to run PylonCapture module on Linux machine (Ubuntu 20.04)

### DISCUSSION THREAD

**Discussion by chiyu1203** [bonsai-rx/bonsai#1679] [Q&A] [answered]
Dear Bonsai community, 
I am writing this to raise an issue about using pylon package on Linux machine. 

I followed this [instruction](https://github.com/orgs/bonsai-rx/discussions/1101) to install Bonsai on Ubuntu 20.04. The installation went relatively well (I only bumped into errors when installing OpenCV: I failed to install OpenCV from **pre-built binaries** but got it working from the **source**). Then I installed pylon viewer 7.1.0 to obtain the 7.1.0 runtime and Pylon package in Bonsai as I did with Windows machines. The pylon viewer can detect and trigger our Basler cameras. However, when I used PylonCapture, Bonsai showed this error and shut down itself.

> Method '<Module>:_getFiberPtrId ()' in assembly '/home/locustvr2/Documents/Bonsai/Packages/Bonsai.Pylon.0.3.0/build/net462/bin/x86/Basler.Pylon.dll' contains native code that cannot be executed by Mono on this platform. The assembly was probably created using C++/CLI.
> 
> System.TypeInitializationException: The type initializer for '<Module>' threw an exception. ---> <CrtImplementationDetails>.ModuleLoadException: The C++ module failed to load.
>  ---> System.MissingMethodException: Method contains unsupported native code assembly:<unknown assembly> type:<unknown type> member:(null)
>   at (wrapper managed-to-native) <Module>._getFiberPtrId()
>   at <Module>.<CrtImplementationDetails>.LanguageSupport._Initialize (<CrtImplementationDetails>.LanguageSupport* ) [0x0001c] in <b20649cf10cc4e098dac1e6794196faf>:0 
>   at <Module>.<CrtImplementationDetails>.LanguageSupport.Initialize (<CrtImplementationDetails>.LanguageSupport* ) [0x00028] in <b20649cf10cc4e098dac1e6794196faf>:0 
>    --- End of inner exception stack trace ---
>   at <Module>.<CrtImplementationDetails>.ThrowModuleLoadException (System.String errorMessage, System.Exception innerException) [0x00007] in <b20649cf10cc4e098dac1e6794196faf>:0 
>   at <Module>.<CrtImplementationDetails>.LanguageSupport.Initialize (<CrtImplementationDetails>.LanguageSupport* ) [0x00041] in <b20649cf10cc4e098dac1e6794196faf>:0 
>   at <Module>..cctor () [0x00008] in <b20649cf10cc4e098dac1e6794196faf>:0 
>    --- End of inner exception stack trace ---
>   at Bonsai.Pylon.PylonCapture+SerialNumberConverter.GetStandardValues (System.ComponentModel.ITypeDescriptorContext context) [0x00000] in <162d37233436490cb440195369186319>:0 
>   at System.Windows.Forms.PropertyGridInternal.GridEntry.get_AcceptedValues () [0x00023] in <25c2bdf9d79a43a599eb46bd15cd77fb>:0 
>   at System.Windows.Forms.PropertyGridInternal.PropertyGridView.UpdateItem (System.Windows.Forms.PropertyGridInternal.GridEntry entry) [0x000b9] in <25c2bdf9d79a43a599eb46bd15cd77fb>:0 
>   at System.Windows.Forms.PropertyGridInternal.PropertyGridView.SelectItem (System.Windows.Forms.PropertyGridInternal.GridEntry oldItem, System.Windows.Forms.PropertyGridInternal.GridEntry newItem) [0x0000d] in <25c2bdf9d79a43a599eb46bd15cd77fb>:0 
>   at (wrapper remoting-invoke-with-check) System.Windows.Forms.PropertyGridInternal.PropertyGridView.SelectItem(System.Windows.Forms.PropertyGridInternal.GridEntry,System.Windows.Forms.PropertyGridInternal.GridEntry)
>   at System.Windows.Forms.PropertyGrid.SelectItemCore (System.Windows.Forms.PropertyGridInternal.GridEntry oldItem, System.Windows.Forms.PropertyGridInternal.GridEntry item) [0x0000e] in <25c2bdf9d79a43a599eb46bd15cd77fb>:0 
>   at System.Windows.Forms.PropertyGrid.set_SelectedGridItem (System.Windows.Forms.GridItem value) [0x0001e] in <25c2bdf9d79a43a599eb46bd15cd77fb>:0 
>   at (wrapper remoting-invoke-with-check) System.Windows.Forms.PropertyGrid.set_SelectedGridItem(System.Windows.Forms.GridItem)
>   at System.Windows.Forms.PropertyGridInternal.PropertyGridView.set_SelectedGridItem (System.Windows.Forms.PropertyGridInternal.GridEntry value) [0x00000] in <25c2bdf9d79a43a599eb46bd15cd77fb>:0 
>   at System.Windows.Forms.PropertyGridInternal.PropertyGridView.OnMouseDown (System.Windows.Forms.MouseEventArgs e) [0x000ad] in <25c2bdf9d79a43a599eb46bd15cd77fb>:0 
>   at System.Windows.Forms.Control.WmLButtonDown (System.Windows.Forms.Message& m) [0x0008b] in <25c2bdf9d79a43a599eb46bd15cd77fb>:0 
>   at System.Windows.Forms.Control.WndProc (System.Windows.Forms.Message& m) [0x001bc] in <25c2bdf9d79a43a599eb46bd15cd77fb>:0 
>   at System.Windows.Forms.ScrollableControl.WndProc (System.Windows.Forms.Message& m) [0x00000] in <25c2bdf9d79a43a599eb46bd15cd77fb>:0 
>   at System.Windows.Forms.Control+ControlWindowTarget.OnMessage (System.Windows.Forms.Message& m) [0x00000] in <25c2bdf9d79a43a599eb46bd15cd77fb>:0 
>   at System.Windows.Forms.Control+ControlNativeWindow.WndProc (System.Windows.Forms.Message& m) [0x0000b] in <25c2bdf9d79a43a599eb46bd15cd77fb>:0 
>   at System.Windows.Forms.NativeWindow.WndProc (System.IntPtr hWnd, System.Windows.Forms.Msg msg, System.IntPtr wParam, System.IntPtr lParam) [0x00085] in <25c2bdf9d79a43a599eb46bd15cd77fb>:0 

Is this some compatibility issue using Bonsai under mono framework to run PylonCapture?
Best,
Chiyu

**Comment by bruno-f-cruz** [comment, marked as answer]
The dlls shipped with the current bonsai package are for Windows (x64 and x86) only from what I can tell from the Pylon website.

**Reply by chiyu1203** [reply]
I see! I dont see pylon 7.1.0 runtime can be downloaded in other OS either....so would this be part of your plan in the next pylon package (to the package compatible in Linux, or there is still a technical issue to overcome?)
